src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -227,6 +238,813 @@ class OMPSerialDirective(OMPRegionDirective, metaclass=abc.ABCMeta):
+                      dependency
+        :type task1: :py:class:`psyclone.psyir.nodes.OMPTaskDirective`
+        :param task2: the OMPTaskDirective node containing node2 as a \
+                      dependency
+        :type task2: :py:class:`psyclone.psyir.nodes.OMPTaskDirective`
+
+        :returns: whether the dependence is going to be handled safely \
+                  according to the OpenMP standard.
+        :rtype: bool
+        '''
+        # Checking the symbol is the same works. If the symbol is not the same
+        # then there's no dependence, so its valid.
+        if node1.symbol != node2.symbol:
+            return True
+        # For structure reference we need to check they access
+        # the same member. If they don't, no dependence so valid.
+        if isinstance(node1, StructureReference):
+            # If either is a StructureReference here they must both be,
+            # as they access the same symbol.
+            member = node1
+            member1 = node2
+
+            # We can't just do == on the Member child, as that
+            # will recurse and check the array indices for any
+            # ArrayMixin children
+
+            # Check the signature of both StructureReference
+            # to see if they are accessing the same data
+            ref0_sig = node1.get_signature_and_indices()[0]
+            ref1_sig = node2.get_signature_and_indices()[0]
+            if ref0_sig != ref1_sig:
+                return True
+
+        # If we have (exactly) Reference objects we filter out
+        # non-matching ones with the symbol check, and matching ones
+        # are always valid since they are simple accesses.
+        if type(node1) is Reference:
+            return True
+
+        # All remaining objects are some sort of Array access
+        array1 = None
+        array2 = None
+        if isinstance(node1, ArrayReference):
+            array1 = node1
+            array2 = node2
+        else:
+            array1 = node1.walk(ArrayMixin)[0]
+            array2 = node2.walk(ArrayMixin)[0]
+        valid = True
+        for i, index in enumerate(array1.indices):
+            if (isinstance(index, Literal) or
+                    isinstance(array2.indices[i], Literal)):
+                valid = valid and self._valid_dependence_literals(
+                            index, array2.indices[i])
+            elif (isinstance(index, Range) or
+                  isinstance(array2.indices[i], Range)):
+                valid = valid and self._valid_dependence_ranges(
+                            array1, array2, i)
+            else:
+                # The only remaining option is that the indices are
+                # References or BinaryOperations
+                valid = valid and self._valid_dependence_ref_binop(
+                            index, array2.indices[i], task1, task2)
+
+        return valid
+
+    def _validate_task_dependencies(self):
```

Comment:
I think we changed the name of this before to this - I'd argue the name is ok so it does validate them (and those that aren't valid become valid through addition).

It could be split but you would end up with some (relatively) meaningless lists being visible in another function, or this function would just call the other (which doesn't change the naming issue here). Once the get_common_ancestor code is refactored away, I don't think this function does that much so I'm not super keen to split it, but if others also feel it should be done I can.